### PR TITLE
updates to work with ActiveRecord 5

### PR DESCRIPTION
**Pull request by kellyredding**
This does the following:

* updates the dependency requirements to only allow AR 5.*.
* updates the migration template to include the AR version per
  Rails's new requirements.
* updates the require autoloaded files script to include the
  appropriate AR files. I found what to update by running
  `bx ruby script/determine_autoloaded_active_record_files.rb`
  with the updated AR 5 dependencies.

# Other Changes

### add new migrate commands for up/down/forward/backward variants

In addition to `ardb migrate`, you now how migrate commands for
specific variants that are available in AR 5.

```
Commands:
  ...
  migrate             # Migrate the configured DB
  migrate-up          # Migrate the configured DB up
  migrate-down        # Migrate the configured DB down
  migrate-forward     # Migrate the configured DB forward
  migrate-backward    # Migrate the configured DB backward
```

The up/down variants take an optional target version:

```
$ ardb migrate-up --help
Usage: ardb migrate-up [options]

Options:
    -t, --target-version VALUE       version to migrate to
        --version
        --help

Description:
  Migrate the configured DB up
```

The forward/backward variants take an optional step value:

```
$ ardb migrate-backward --help
Usage: ardb migrate-backward [options]

Options:
    -s, --steps VALUE                number of migrations to migrate
        --version
        --help

Description:
  Migrate the configured DB backward
```

### rework commands to configure their own name/summary

This makes it so you just add commands to the command set and
don't have to specify the command name.

After adding the new migration commands, it felt weird to configure
the summary and help message on the command object but configure
the commmand name separately. Also the command name in the help
message had to be kept in sync with the configured command name.

This reworks so that the command name is configured like the
summary and the help message is now a generated template on the
ValidCommand mix-in.

This also adds tests for the new migration commands which I forgot
to add in the previous commit where I added the commands
themselves.

### update commands for better UX around the database existing

The updates the CLI commands so that they detect whether the
database exists or not and output friendly error messages if the
database is in a bad state for the command being attempted:

```
$ ardb connect
error: database "some_database" does not exist.
$ ardb drop
error: database "some_database" does not exist.
$ ardb migrate
error: database "some_database" does not exist.
$ ardb create
created postgresql db `some_database`
$ ardb create
error: database "some_database" already exists.
```

Previously, the commands would spit out raw exception style
output with raw attempted SQL and/or backtraces. This isn't
really friendly or necessary when the root cause of the problem
is this obvious.

### upgrade to cli.rb v1.1.0; update to latest Assert

This gets on the latest versions to keep up with the latest stuff.
This also restricts the ruby version to 2.4+ in the gemspec.

### validate the configured database name has no non-word characters

This updates the validation logic to add this validation. I ran
into problems running commands on databases with non-word
characters in their name. Specifically, I couldn't drop a postgres
db with non-word characters in the name.

This validation prevents/protects the user from getting in a bad
state b/c of the database name.

### CODE REVIEW

**Comment by kellyredding** [comment]
@jcredding note: with all of these changes, I am able to add in in `config/db.rb` file to a Rails app and use Ardb CLI to manage the database and the migrations. I am also able to run system tests against the db in our style in the Rails app. 💥 

Here's the Rails compatible `config/db.rb` file I used:

```ruby
require_relative "./environment"
require "ardb"

# This Ardb configuration matches Rails's handling.
Ardb.configure do |c|
  rails_db_config = Rails.application.config_for("database")
  c.root_path     = Rails.root
  c.logger        = Rails.logger
  c.schema_format = Rails.application.config.active_record.schema_format || :ruby
  c.adapter       = rails_db_config["adapter"]
  c.host          = rails_db_config["host"]
  c.port          = rails_db_config["port"]
  c.username      = rails_db_config["username"]
  c.password      = rails_db_config["password"]
  c.database      = rails_db_config["database"]
  c.encoding      = rails_db_config["encoding"]
  c.min_messages  = rails_db_config["min_messages"]

  c.migrations_path = "db/migrate"
  c.schema_path = "db/schema"
end
```

File: test/unit/adapter/base_tests.rb [hunk 1]
```diff
@@ -115,41 +115,19 @@ class ConnectDbTests < UnitTests
...
-      @orig_migrate_version_env_var = ENV["MIGRATE_VERSION"]
-      @orig_migrate_query_env_var   = ENV["MIGRATE_QUIET"]
-
-      ENV["MIGRATE_VERSION"] = Factory.integer.to_s if Factory.boolean
-      ENV["MIGRATE_QUIET"]   = Factory.boolean.to_s if Factory.boolean
-
-      @migrator_called_with = []
-      Assert.stub(ActiveRecord::Migrator, :migrate) do |*args|
-        @migrator_called_with = args
+      skip
```
Comment (jcredding) [inline test/unit/adapter/base_tests.rb]: @kellyredding - Are you planning on addressing this in a future commit or did you want to handle it here?

Comment (kellyredding) [reply test/unit/adapter/base_tests.rb]: No, forgot about this so thanks - I need to deal with it here - stay tuned.

Comment (kellyredding) [reply test/unit/adapter/base_tests.rb]: Ok, I've added proper tests for the new migrate commands - please look again thanks!

**Comment by jcredding** [comment]
Thought I wrote this yesterday but thanks for figuring all this stuff out. I really like the new migration stuff and the changes to the commands.

**Pull request merged**